src/plugins/hyperbee/hyperbee.cpp

```diff
@@ -0,0 +1,1329 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************
+ * This plugin is part of the master's thesis of Klaus-Günther Schmidt,    *
+ * student of FAU Erlangen-Nürnberg (DE) in cooperation with Politecnico   *
+ * di Milano (IT) during 2020/2021.                                        *
+ * The main goal of this plugin is to efficiently dump executable pages    *
+ * which afterwards could be processed by analysis tools to detect         *
+ * malware. The goal is reached by installing several traps:               *
+ * 1. mmAccessFaultTrap: Whenever a MmAccessFault is called (to commit     *
+ *    virtual memory) the regarding virtual address is saved. Additionally *
+ *    the second trap is set up:                                           *
+ * 2. When MmAccessFault returns to the original code it provides the      *
+ *    physical address the virtual address was assigned to. This           *
+ *    information is saved as well and used to set up a third trap:        *
+ * 3. Whenever instructions are fetched from the physical address (frame)  *
+ *    to be executed, this trap dumps the memory of the frame or even the  *
+ *    whole vad node.                                                      *
+ *    To be aware of further executions of this page but prevent duplicate *
+ *    dumps, this execution trap is replaced by a fourth, write trap.      *
+ * 4. The write trap is reacting on any writes at the monitored memory and *
+ *    replaces itself with an executable trap (3). By continuously         *
+ *    swapping trap 3 and 4 memory a dump of the same memory area is only  *
+ *    performed, if the memory was modified in the mean time.              *
+ **************************************************************************/
+
+#include <glib.h>
+#include <cinttypes>
+#include <sys/stat.h>
+#include <libvmi/libvmi.h>
+#include <string>
+#include <vector>
+#include <cstdio>
+#include <cstring>
+
+#include <plugins/filesystem.hpp>
+#include <libdrakvuf/json-util.h>
+#include "hyperbee.h"
+#include "plugins/output_format.h"
+#include "private.h"
+
+/**
+ * The default string, for incomplete entries.
+ */
+static char missing_data[] = "(null)";
+/**
+ * The string used for vad nodes, that don't belong to mapped files, but to dynamically allocated memory
+ */
+static char alloc_memory[] = "(no-mapped-file)";
+/**
+ * array with the two different analysis modes as strings
+ */
+static char file_extensions[][5] = {"page", "vad"};
```

Comment:
1) this is kinda cursed
2) but it's used only once in the code. please inline it in there.

Reply:
What do you mean with inline? Just make it a local variable?

Reply:
In `set_dump_paths` we currently have:
```cpp
    if (asprintf(&dump_metadata->dump_file, "%s/%s.%s", dump_dir, dump_metadata->file_stem, file_extensions[dump_vad])
```
I propose:
```cpp
    auto file_extension = dump_vad ? "vad" : "page";
    if (asprintf(&dump_metadata->dump_file, "%s/%s.%s", dump_dir, dump_metadata->file_stem, file_extension)
```

Reply:
Yeah, thats way better!